### DISCUSSION TITLE
High memory consumption!

### DISCUSSION THREAD

**Discussion by serzhiio** [DioxusLabs/dioxus#2195] [General]
After recent update i noticed significant increase of browser memory consumption.
With early 0.5 ot was less than 100mb and now it is about 3G!!!

**Comment by serzhiio** [comment]
My system UI is working 24/7 and memory consumption was flat all time, until this day it starts from 100mb and after some time increasing.

**Comment by serzhiio** [comment]
![Снимок экрана от 2024-03-30 17-32-35](https://github.com/DioxusLabs/dioxus/assets/10449591/b8887511-6430-4b11-8adb-36186952fd0b)

**Comment by ealmloff** [comment]
Are you creating any signals outside of hooks? Signal::new will be owned by the current scope which means if you call `Signal::new` in a loop in a component, it will not be dropped (unless you manually drop it) until the component it was created in is dropped

**Comment by serzhiio** [comment]
Almost all signals are global.
UPD: Double checked everything, i have no Signal::new() invocations at all. Only use_signal() in components at the top.

**Reply by jkelleyrtp** [reply]
Gonna try and diagnose and put out a fix for this ASAP - should be this week if we can figure it out.

**Comment by serzhiio** [comment]
I believe early 0.5 alpha was not affected by this, but not 100% sure
Still investigating my own code, but the only updates i did are aligning everything with new Signals.
UPD: No, early 0.5 also shows same memory consumption. Unfortunately i can't revert to 0.4.3.

**Comment by jkelleyrtp** [comment]
Is this code open source at all? Or can you post some of the signal bits?

I can't replicate this in chrome - the memory usage stays flat when using global signals.

Also in your screenshot the memory usage is 164,298,752, which is roughly 160mb. That's not 3gb so maybe it's some other site?

We have two options here:
- a minimum viable replication
- or you have to do some debugging with https://github.com/rustwasm/wasm-tracing-allocator 

---

Did some debugging and memory usage does grow with hotreload - and we have hotreload enabled by default in debug mode now. 

Perhaps you're editing files locally? We leak memory to support hotreload - it's just designed that way and we really don't have a desire to change it.

Can you verify this still happens in release?

**Reply by serzhiio** [reply]
>Is this code open source at all? Or can you post some of the signal bits?

No, it is my private project, not exposed to public.

> We have two options here:
> 
> ->a minimum viable replication
> ->or you have to do some debugging with https://github.com/rustwasm/wasm-tracing-allocator

I am trying `wasm-tracing-allocator` but it doesnt show traces only errors.

>Perhaps you're editing files locally? We leak memory to support hotreload - it's just designed that way and we really don't have a desire to change it.

No it is production code, built with `dx build --release`

P.S.: After migration to 0.5 i am doing a lot of cloning inside components.
![Снимок экрана от 2024-04-03 14-09-28](https://github.com/DioxusLabs/dioxus/assets/10449591/5dc8f4a7-4753-47f4-8afb-d4222c3f4248)

**Reply by jkelleyrtp** [reply]
The tracing allocator uses errors to get stack traces - try it out in firefox and if you dig deep enough into the errors you should get the locations where the allocations are happening. For me, the dominators are in the hotreload init code.

See more: https://github.com/rustwasm/wasm-tracing-allocator/issues/6

**Comment by serzhiio** [comment]
Updated gist without hot reload traces.
[link](https://gist.github.com/serzhiio/8bcc7a571d4739fa342818cfb32d318d)

**Comment by serzhiio** [comment]
After i commented all components in code and started to uncomment each part separately i was finally able to isolate memory consumption source, this is event handlers:
```
        let subs = if let Some(subs) = props.subs.as_ref() {
            let tmp = subs.map.values().map(|v| v.clone()).collect::<Vec<_>>();
            let subs = tmp.into_iter().filter(|s| !s.is_empty()).map(|sub| {
                let mkt_inst = *sub.market_instrument();
                let streams = sub.get_streams().iter().copied().map(|s| {
                    let stream = *s.stream();
                    rsx!(
                        div{
                            class:"flex-none px-1 bg-gray-800 uppercase font-bold text-[10px] text-blue-300 hover:bg-red-900 hover:text-white cursor-pointer",
--->                      /*onclick: move |_| {
                                signal.send(Signal::MpubSubscriptionUpdate(props.id, mkt_inst, stream, StreamState::ToRemove))
                            },*/
                            "{s.stream().as_first_letter()}"
                        }
                    )
                });
                let excl = if sub.is_exclusive() { "!" } else { "~" };
                // class: "font-bold text-purple-400"
                let excl_styles = if sub.is_exclusive() { " font-bold text-amber-500" } else { "" };
                rsx!(
                    div{class:"flex items-start rounded-md bg-gray-800 text-xs px-1 mb-0.5 mr-0.5",
                        div{class:"flex-none uppercase font-bold text-[10px] pr-0.5 text-gray-300",
                            "{sub.instrument()}"
                        }
                        {streams}
                        div{class:"flex-none text-[10px]{excl_styles} hover:bg-red-900 hover:text-white cursor-pointer px-0.5",
--->                      /*onclick: move |_| {
                                signal.send(Signal::MpubSubscriptionExclusivity(props.id, *sub.market_instrument(), !sub.is_exclusive()))
                            },*/
                            "{excl}"
                        }
                    }
                )
            });
            Some(rsx!({ subs }))
        } else {
            None
        };

rsx!(subs)
```

**Reply by serzhiio** [reply]
In the 0.4.3 it was possible to use a ref inside event handler.

**Comment by jkelleyrtp** [comment]
We tracked in down and it's been fixed for 0.5.1 - thanks for the allocator traces - that was exactly it!

**Reply by serzhiio** [reply]
I am glad it helped! This patch definetely did it.
But memory consumption compared to 0.4.3 is a bit higher, is it expected?

**Comment by serzhiio** [comment]
And it is still looks like there is some minor leak exists. 
I am not sure, but memory consumed is slowly increasing.

**Comment by serzhiio** [comment]
Finally it is gone, i hope :)
